core/src/jsMain/kotlin/Peripheral.kt

```diff
@@ -128,53 +133,105 @@ public class JsPeripheral internal constructor(
-        job.cancelAndJoinChildren()
-        connectJob = null
-        disconnectGatt()
+        disconnectSync(cancelConnectJob = true)
+    }
+
+    private suspend fun connectSync() {
+        val job = connectJob ?: connectAsync().apply {
+            connectJob = this
+            invokeOnCompletion { connectJob = null }
+        }
+        job.await()
+    }
+
+    private suspend fun disconnectSync(
+        cancelConnectJob: Boolean
+    ) {
+        val job = disconnectJob ?: disconnectAsync(cancelConnectJob).apply {
+            disconnectJob = this
+            invokeOnCompletion { disconnectJob = null }
+        }
+        job.await()
+    }
+
+    private suspend fun openConnection() {
+        // Guard if already connected do nothing
+        if (_state.value is State.Connected) return
+
+        logger.info { message = "Connecting" }
+        // Wait until fully disconnected before proceeding
+        _state.first { it is State.Disconnected }
+
+        _state.value = State.Connecting.Bluetooth
+        registerDisconnectedListener()
+        gatt.connect().await() // todo: Catch appropriate exception to emit State.Rejected.
+
+        _state.value = State.Connecting.Services
+        discoverServices()
+        onServicesDiscovered(ServicesDiscoveredPeripheral(this@JsPeripheral))
+
+        _state.value = State.Connecting.Observes
+        logger.verbose { message = "Configuring characteristic observations" }
+        observers.onConnected()
+
+        logger.info { message = "Connected" }
+        _state.value = State.Connected
     }
 
-    private fun disconnectGatt() {
-        _state.value = State.Disconnecting
-        bluetoothDevice.gatt?.disconnect()
-        // Avoid trampling existing `Disconnected` state (and its properties) by only updating if not already `Disconnected`.
-        _state.update { previous -> previous as? State.Disconnected ?: State.Disconnected() }
+    private suspend fun closeConnection() {
+        // Guard if already disconnecting/disconnected do nothing
+        if (_state.value is State.Disconnecting || _state.value is State.Disconnected) return
+
+        try {
+            _state.value = State.Disconnecting
+            unregisterDisconnectedListener()
+            stopAllObservations()
+            bluetoothDevice.gatt?.disconnect()
+            // FIXME: https://github.com/JuulLabs/kable/issues/251
+            delay(DISCONNECTION_DELAY)
+        } finally {
+            _state.value = State.Disconnected()
+        }
+    }
+
+    private fun finalCleanup() {
+        try {
+            unregisterDisconnectedListener()
+            clearObservationsWithoutStopping()
+            bluetoothDevice.gatt?.disconnect()
+        } finally {
+            // Avoid trampling existing `Disconnected` state (and its properties) by only updating if not already `Disconnected`.
+            _state.update { previous -> previous as? State.Disconnected ?: State.Disconnected() }
```

Comment:
From what I can tell of Web Bluetooth, we don't get any status codes from the disconnected callback, so `Disconnected` on JS will never carry a status. In other words, I realize now that we don't need to do the `update` dance on JS, because we'll never trample an out-of-band disconnect event. We can just do `_status.value = Disconnected()`.

Reply:
Good catch, will remove this.